agent/api/appnet/client_linux.go

```diff
@@ -0,0 +1,100 @@
+//go:build linux
+
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package appnet
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/aws/amazon-ecs-agent/agent/api/task"
+	"github.com/aws/amazon-ecs-agent/agent/logger"
+	"github.com/aws/amazon-ecs-agent/agent/logger/field"
+	"github.com/aws/amazon-ecs-agent/agent/utils/retry"
+	prometheus "github.com/prometheus/client_model/go"
+)
+
+const (
+	unixNetworkName   = "unix"
+	httpRequestPrefix = "http://" + unixNetworkName
+	statsUrl          = httpRequestPrefix + "/stats/prometheus?usedonly&filter=metrics_extension&delta"
+	drainUrl          = httpRequestPrefix + "/drain_listeners?inboundonly"
+)
+
+type appnetClientCtxKey int
+
+const (
+	udsAddressKey appnetClientCtxKey = iota
+)
+
+var (
+	// Injection point for UTs
+	performAppnetRequest     = doPerformAppnetRequest
+	oneSecondBackoffNoJitter = retry.NewExponentialBackoff(time.Second, time.Second, 0, 1)
+)
+
+func udsDialContext(ctx context.Context, _, _ string) (net.Conn, error) {
+	udsPath, ok := ctx.Value(udsAddressKey).(string)
+	if !ok {
+		return nil, fmt.Errorf("appnet client: Path to appnet admin socket was not a string")
+	}
+	if udsPath == "" {
+		return nil, fmt.Errorf("appnet client: Path to appnet admin socket was blank")
+	}
+	return net.Dial(unixNetworkName, udsPath)
+}
+
+var udsHttpClient = http.Client{
+	Transport: &http.Transport{
+		DialContext: udsDialContext,
+	},
```

Comment:
We would need to add some default retry strategy to the client. But we can take that up separately.

Reply:
I implemented a basic retry for Drain API, which should suffice. 

For GetStats, it's not very clear at this time what would be a good retry strategy. The issue with retrying this API is that if the request does go through, the delta will be reset, so retrying will only get the "next" snapshot, which might not be what we want.